File: week2/seven.cpp
#include <vector>
#include <string>
#include <fstream>
#include <sstream>
#include <iostream>
#include <unordered_map>
#include <unordered_set>
#include <regex>
#include <algorithm>
using namespace std;


int main() {
    string word;
    unordered_set<string> stops;
    unordered_map<string, int> freqs;
    // read stop words
    ifstream stopfile("../stop_words.txt");
    while (getline(stopfile, word, ','))    stops.insert(word);
    // read input and normalize
    ifstream file("../pride-and-prejudice.txt");
    string data = string(istreambuf_iterator<char>(file), istreambuf_iterator<char>());
    for (char &c : data)    c = tolower(c);
    // use regular expresion to fetch words
    regex rg("[a-z]{2,}");
    for (auto it = sregex_iterator(data.begin(), data.end(), rg); it != sregex_iterator(); it++) {
        string word = (*it).str();
        if (!stops.count(word))  freqs[word]++;
    }
    // sort by freq
    vector<pair<string, int>> res(freqs.begin(), freqs.end());
    sort(res.begin(), res.end(), [](pair<string, int> &a, pair<string, int> &b) {
        return a.second > b.second;});
    // print result
    for (int i = 0; i < 25; i++)  cout << res[i].first << "  -  " << res[i].second << endl; 
}

File: week2/Seven.go
package main

import (
	"fmt"
	"io/ioutil"
	"regexp"
	"sort"
	"strings"
)

func main() {
	path := "../pride-and-prejudice.txt"
	data, _ := ioutil.ReadFile(path)

	reg := regexp.MustCompile("[a-z]{2,}")
	words := reg.FindAllString(strings.ToLower(string(data)), -1)

	pathStop := "../stop_words.txt"
	data, _ = ioutil.ReadFile(pathStop)

	stopWrods := strings.Split(string(data), ",")
	stops := make(map[string]bool)
	for _, word := range stopWrods {
		stops[word] = true
	}

	freqs := make(map[string]int)
	for _, word := range words {
		if !stops[word] {
			freqs[word]++
		}
	}

	type kv struct {
		Key   string
		Value int
	}

	var ss []kv
	for k, v := range freqs {
		ss = append(ss, kv{k, v})
	}

	sort.Slice(ss, func(i, j int) bool {
		return ss[i].Value > ss[j].Value
	})

	for i := 0; i < 25; i++ {
		fmt.Println(ss[i].Key, " - ", ss[i].Value)
	}

}
